### ISSUE TITLE
SC2126 not always correct

### ISSUE THREAD

**Issue by Nachtfeuer** [koalaman/shellcheck#741]
I do a lot of bash scripting - currently - for a ci/cd pipeline for one project but I'm probably not as deep in bash scripting like you, however ... following problem:
```
PROMPT="$(basename "$0") :: "

result=$(docker ps -a | grep DOES_NOT_EXIST | wc -l)
error_code=$?
echo "${PROMPT}result=${result}, error_code=${error_code}"

result=$(docker ps -a | grep -c DOES_NOT_EXIST)
error_code=$?
echo "${PROMPT}result=${result}, error_code=${error_code}"
```

The output of this script code is following

```
foo :: result=0, error_code=0
foo :: result=0, error_code=1
```

As you see ... the second variant with grep only fails. Obviously the count mechanism does work fine only when you have at least one entry. From this the "wc -l" variant works better except you can tell me how to solve the mentioned problem.

**Comment by brother** [comment]
the first erro_code will show the exit code of wc -l and the second one shows the exit code of grep. grep will fail and thus show 1. wc -l will never fail and always have 0 as exit code.

brother ~$ wc -l
^C
brother ~$ echo $?
130
brother ~$ echo "hello" | grep A | wc -l 
0
brother ~$ echo $?
0
brother ~$ echo "hello" | grep -c A
0
brother ~$ echo $?
1

what is the actual problem you are trying to solve. The example code is unclear.

**Comment by Nachtfeuer** [comment]
Let me put it more simple (without Docker)
```
echo $'first\nsecond'| grep -c first   => stdout: 1, exit code:0
echo $'first\nsecond'| grep -c nobody  => stdout: 0, exit code:1
```

When I'm writing test cases (bats) then using the "grep -c" variant make the test fail because the grep command fails. But the intention was different - I just wanted to know the **count**. However a grep that has no result (0) additional exits with 1 as you see. That's why "wc -l" works better when just the count is of interest.

**Comment by Nachtfeuer** [comment]
To make it complete
```
tlehmann /tmp $ cat foo.sh 
#!/bin/bash
count=$(echo $'first\nsecond'| grep first|wc -l)
echo "count=$count"
tlehmann /tmp $ shellcheck foo.sh

In foo.sh line 2:
count=$(echo $'first\nsecond'| grep first|wc -l)
                               ^-- SC2126: Consider using grep -c instead of grep|wc.
```

Now try to execute following:
```
$ cat foo.sh 
#!/bin/bash
set -e
set -o pipefail

count=$(echo $'first\nsecond'| grep -c unknown)
echo "count=$count"
```
You will **NOT** see "count=0". It does work with "wc -l".
Again: that's why I see that the warning is sometimes ... questionable ...
Except you can say me how to work with "grep -c" without this problem.

**Comment by brother** [comment]
Nothing is perfect when it comes to recommendations and suggestions. The warning tells you to consider using another form. If you have considered that (I think you have) you put a disable statement before it and shellcheck won't bother you about it.

https://github.com/koalaman/shellcheck/wiki/SC2126 has a pretty clear note (and maybe the warning message can be made clearer - I have no suggestion though): "This is purely a stylistic issue. grep can count lines without piping to wc."

Your definition of "works" is different than mine, mostly because you focus on your small context. Both grep and wc are working, not to your liking.

Not wanting grep to exit with 1 when it has no result is by all means strange.

**Comment by kurahaupo** [comment]
On Fri, 30 Sep 2016, Thomas Lehmann wrote:
> Now try to execute following:
> ```
> $ cat foo.sh
> #!/bin/bash
> set -e
> set -o pipefail
>
> count=$(echo $'first\nsecond'| grep -c unknown)
> echo "count=$count"
> ```
> You will **NOT** see "count=0".

True. But we can debate which part of the code is wrong.

> It does work with "wc -l".

Are you sure about that?

When I try this:
```
set -e -o pipefail
count=$(echo $'first\nsecond'| grep -c unknown | wc -l)
```
it fails.

(When you use `set -o pipefail` in combination with `set -e`, if ANY
component of the pipeline fails, the shell will see this as a command
failure and will abort; however this may vary between different versions of
Bash, or with different shells.)

As a separate point, some would argue that `set -e` is the real problem.
Although some shell guides (and even some training courses) suggest always
including this line, the more sophisticated guides recommend avoiding it,
precisely because `set -e` has a bunch of corner cases where valid code
(like this) breaks, and failing code still doesn't get caught.

In short: `set -e` and `set -u` should be treated as linting tools, not for
use in production code.

> Again: that's why I see that the warning is sometimes ... questionable ...
> Except you can say me how to work with "grep -c" without this problem.

You could also write:
```
count=$( somecmd | grep -c somepattern ) || :
```
which would take rather less space that the comment to disable shellcheck.

Or even
```
count=$( somecmd > >( grep -c somepattern ) )
```
which (since you've used pipefail) will trap when somecmd fails, but won't
trap when grep fails -- with or without pipefail.

**Comment by koalaman** [comment]
I'd agree with kurahaupo, this is a good stylistic suggestion that rather exposes an existing issue with the code. 

Correctness relies on the fact that `wc` happens to mask the exit status of grep when `pipefail` is disabled. I'm guessing this was through luck more than by design. 

Rather than relying on this somewhat obscure bit of shell knowledge, it's better to explicitly use `|| true` or equivalent on lines where you don't care about the exit status when using `set -e`  (or to [not use `set -e` at all](http://mywiki.wooledge.org/BashFAQ/105)).

I'd agree that this is far from obvious though, and I've updated the wiki to mention this case. Thanks!
